src/spikeinterface/sortingcomponents/benchmark/benchmark_matching.py

```diff
@@ -16,101 +17,483 @@
+            else:
+                method_kwargs["waveform_extractor"] = we
+        return methods_kwargs
 
-    def run(self):
-        t_start = time.time()
-        self.spikes = find_spikes_from_templates(
-            self.recording, method=self.method, method_kwargs=self.method_kwargs, **self.job_kwargs
+    def run_matching_misclassed(
+        self, fraction_misclassed, min_similarity=-1, seed=0, we_kwargs=None, template_mode="median"
+    ):
+        """Run template matching with a given fraction of misclassified spikes.
+
+        Parameters
+        ----------
+        fraction_misclassed: float
+            The fraction of misclassified spikes.
+        min_similarity: float
+            The minimum cosine similarity between templates to be considered similar. (Default: -1)
+        seed: int
+            Random seed. (Default: 0)
+        we_kwargs: dict
+            A dictionary of keyword arguments for the WaveformExtractor.
+        template_mode: {'mean' | 'median' | 'std'}
+            The mode to use to extract templates from the WaveformExtractor. (Default: 'median')
+
+        Returns
+        -------
+        sortings: dict
+            A dictionary that maps method --> NumpySorting.
+        gt_sorting: NumpySorting
+            The ground-truth sorting used for template matching (with misclassified spike trains).
+        """
+        try:
+            assert 0 <= fraction_misclassed <= 1
+        except AssertionError:
+            raise ValueError("'fraction_misclassed' must be between 0 and 1.")
+        try:
+            assert -1 <= min_similarity <= 1
+        except AssertionError:
+            raise ValueError("'min_similarity' must be between -1 and 1.")
+        if we_kwargs is None:
+            we_kwargs = {}
+        we_kwargs.update(dict(seed=seed, overwrite=True, load_if_exists=False, **self.job_kwargs))
+        rng = np.random.default_rng(seed)
+
+        # Randomly misclass spike trains
+        spike_time_indices, labels = [], []
+        for unit_index, unit_id in enumerate(self.we.unit_ids):
+            unit_sorting = self.gt_sorting.get_unit_spike_train(unit_id=unit_id)
+            unit_similarity = self.similarity[unit_index, :]
+            unit_similarity[unit_index] = min_similarity - 1  # skip self
+            similar_unit_ids = self.we.unit_ids[unit_similarity >= min_similarity]
+            at_least_one_similar_unit = len(similar_unit_ids)
+            num_spikes = int(len(unit_sorting) * fraction_misclassed)
+
+            unit_misclass_idx = rng.choice(np.arange(len(unit_sorting)), size=num_spikes, replace=False)
+            for i, spike in enumerate(unit_sorting):
+                spike_time_indices.append(spike)
+                if i in unit_misclass_idx and at_least_one_similar_unit:
+                    alt_id = rng.choice(similar_unit_ids)
+                    labels.append(alt_id)
+                else:
+                    labels.append(unit_id)
```

Comment:
I have the feeeling that this loop could be numpy vectorized, no ?
looping over spikes can be super long no ?